pkg/framework/resource/parse.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/util/openapi"
+)
+
+// Parser is an interface for an object that can be used to discover resources from an API server and parse them into indexed data structures.
+type Parser interface {
+	Resources() (Resources, error)
+}
+
+// parser is an object type that can be used to discover resources from an API server and parse them into indexed data structures.
+type parser struct {
+	resources  openapi.Resources
+	discovery  discovery.DiscoveryInterface
+	apiGroup   string
+	apiVersion string
+}
+
+// NewParser populates the fields of and returns a new parser object.
+func NewParser(resources openapi.Resources, discovery discovery.DiscoveryInterface, apiGroup string, apiVersion string) Parser {
+	return &parser{
+		resources,
+		discovery,
+		apiGroup,
+		apiVersion,
+	}
+}
+
+// Resources discovers and indexes resources from the API server.
+// It returns a Resources object: a map of resource names and their associated Resource objects, ordered by preference as reported by the server.
+func (p *parser) Resources() (Resources, error) {
+	gvs, err := p.discovery.ServerResources()
+	if err != nil {
+		return nil, err
+	}
+	resources, byGVR := p.indexResources(gvs)
+	err = p.attachSubResources(gvs, resources, byGVR)
+	return resources, err
+}
+
+// subResource returns a resource name, subresource name pair, and true if the resource is a subresource.
+// It returns a resource name, an empty string, and false if the resource is not a subresource.
+func (*parser) subResource(resource *v1.APIResource) (string, string, bool) {
```

Comment:
since the meaning of these return parameters can't be inferred by their type, you should maybe name them.